internal/gps/checks.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gps
+
+import (
+	"github.com/golang/dep/internal/gps/pkgtree"
+)
+
+func FindIneffectualConstraints(manifest Manifest, packageTree pkgtree.PackageTree) []ProjectRoot {
+	var stdLibFn func(string) bool
+	// will track return value here
+	ineffectuals := make([]ProjectRoot, 0, 10)
+
+	// flatten list of actual imports which should be checked against
+	reachmap, _ := packageTree.ToReachMap(true, true, false, nil /*ignores*/)
+	reach := reachmap.FlattenFn(stdLibFn)
+	imports := make(map[string]bool)
+	for _, imp := range reach {
+		imports[imp] = true
+	}
+
+	// if manifest is a RootManifest, use requires and ignores
+	if rootManifest := manifest.(RootManifest); rootManifest != nil {
+		// add required packages
+		for projectRoot, _ := range rootManifest.RequiredPackages() {
+			imports[projectRoot] = true
+		}
+
+		// check that ignores actually refer to packages we're importing
+		for ignore, _ := range rootManifest.IgnoredPackages() {
+			if _, found := imports[ignore]; !found {
+				ineffectuals = append(ineffectuals, ProjectRoot(ignore))
+			}
+		}
+
+		// TODO: Should we remove ignores from the list? Ignores should not be
```

Comment:
interesting. this is a bit tricky, but yes, i think we have to do it.

the trick is that, because constraints are applied strictly to the project root, but ignores are package-oriented, we have to keep track of whether ignores eliminate _all_ imports from within a given project root, not just _some_. only if all are eliminated should we treat constraints on that project as ineffectual.